### PR TITLE
8298720: Insufficient error handling when CodeBuffer is exhausted

### PR DESCRIPTION
**Pull request by TobiHartmann**
This patch fixes various places in C1 and C2 on Aarch64 and RISC-V that miss proper error handling when the code buffer is exhausted, leading to crashes. Similar but incomplete patches went in with [JDK-8130309](https://bugs.openjdk.org/browse/JDK-8130309), [JDK-8248411](https://bugs.openjdk.org/browse/JDK-8248411) and [JDK-8272094](https://bugs.openjdk.org/browse/JDK-8272094) in the past. 

These issues are extremely hard to reproduce, even with the `-XX:+StressCodeBuffers` option, because code buffer expansion needs to fail at the exact moment when a specific (unhandled) instruction is emitted. Even with the stress option, we expand the code buffer such that multiple instructions will fit and in addition, chances are high that we simply bail out from compilation before emitting the problematic instruction. I attached a patch to [JDK-8298720](https://bugs.openjdk.org/browse/JDK-8298720), that makes `-XX:+StressCodeBuffers` randomized and more aggressive. With that, I can reproduce the issue reliably but it's extremely slow and therefore not well suited for integration.

I now went over all usages of `CodeBuffer::expand` to make sure that we have proper error handling in place and found some remaining issues in JVMCI code. I filed [JDK-8299570](https://bugs.openjdk.org/browse/JDK-8299570) to address them.

I would need help to test the RISC-V specific changes.

Thanks,
Tobias

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298720](https://bugs.openjdk.org/browse/JDK-8298720): Insufficient error handling when CodeBuffer is exhausted

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11839/head:pull/11839` \
`$ git checkout pull/11839`

Update a local copy of the PR: \
`$ git checkout pull/11839` \
`$ git pull https://git.openjdk.org/jdk pull/11839/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11839`

View PR using the GUI difftool: \
`$ git pr show -t 11839`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11839.diff">https://git.openjdk.org/jdk/pull/11839.diff</a>

</details>

### CODE REVIEW

**Comment by RealFYang** [comment]
Hi! Thanks for handling RISC-V at the same time. I can help arrange for some necessary tests on linux-riscv64 for those changes.

**Comment by TobiHartmann** [comment]
Thanks @RealFYang! It's mostly about verifying that I did not miss any label in the `reset_labels` calls. You should be able to hardcode the bailout to always true and check if we don't hit any assert.

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp [hunk 1]
```diff
@@ -1194,13 +1194,17 @@ void LIR_Assembler::emit_alloc_obj(LIR_OpAllocObj* op) {
...
-                     *op->stub()->entry());
+  address tpc = __ allocate_object(op->obj()->as_register(),
+                                   op->tmp1()->as_register(),
+                                   op->tmp2()->as_register(),
+                                   op->header_size(),
+                                   op->object_size(),
+                                   op->klass()->as_register(),
+                                   *op->stub()->entry());
+  if (tpc == nullptr) {
+    bailout("trampoline stub overflow");
```
Comment (vnkozlov) [inline src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:1205]: The message is confusing. How about `"no space for trampoline stub"` used in `c1_LIRAssembler_ppc.cpp`.

**Review by vnkozlov — COMMENTED** [review]
My main question is: can we do bailout in `MacroAssembler::trampoline_call()` instead of later in all over places?
In `c1_LIRAssembler_ppc.cpp` it bailout in emit_trampoline_stub_for_call().

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp (hunk 1 above)
Comment (TobiHartmann) [reply src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp]: Makes sense. Fixed.

**Comment by RealFYang** [comment]
> Thanks @RealFYang! It's mostly about verifying that I did not miss any label in the `reset_labels` calls. You should be able to hardcode the bailout to always true and check if we don't hit any assert.

Hi, fastdebug bootcycle is fine on linux-riscv64 platform.
And If I hardcoded the bailout to always true, I got following error when I do native fastdebug build on both aarch64 and riscv64 platforms:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/realfyang/openjdk-jdk/src/hotspot/share/opto/output.cpp:3241), pid=2565744, tid=2566009
#  assert(!C->failing()) failed: Must not have pending failure. Reason is: CodeCache is full
#
# JRE version: OpenJDK Runtime Environment (21.0) (fastdebug build 21-internal-adhoc.realfyang.openjdk-jdk)
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 21-internal-adhoc.realfyang.openjdk-jdk, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# V  [libjvm.so+0x14fa71c]  PhaseOutput::scratch_emit_size(Node const*)+0x314
```

**Review by RealFYang — APPROVED** [review]
Looks good to me.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by TobiHartmann** [comment]
Thanks, Vladimir!

**Pull request closed without merging**